src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -0,0 +1,358 @@
+                    (uint)OLECMDEXECOPT.OLECMDEXECOPT_DODEFAULT,
+                    null));
+
+            KeybindingsResetLogger.Log("KeybindingsReset");
+
+            _workspace.Options = _workspace.Options.WithChangedOption(KeybindingResetOptions.NeedsReset, false);
+        }
+
+        private void OpenExtensionsHyperlink()
+        {
+            ThisCanBeCalledOnAnyThread();
+            Process.Start(KeybindingsFwLink);
+
+            KeybindingsResetLogger.Log("ExtensionsLink");
+            _workspace.Options = _workspace.Options.WithChangedOption(KeybindingResetOptions.NeedsReset, false);
+        }
+
+        private void NeverShowAgain()
+        {
+            AssertIsForeground();
+
+            _workspace.Options = _workspace.Options.WithChangedOption(KeybindingResetOptions.NeverShowAgain, true);
+            KeybindingsResetLogger.Log("NeverShowAgain");
+
+            // The only external references to this object are as callbacks, which are removed by the dispose method.
+            Dispose();
+        }
+
+        private void InfoBarClose()
+        {
+            AssertIsForeground();
+            _infoBarOpen = false;
+        }
+
+        public int QueryStatus(ref Guid pguidCmdGroup, uint cCmds, OLECMD[] prgCmds, IntPtr pCmdText)
+        {
+            // We don't care about query status, only when the command is actually executed
+            return (int)OLE.Interop.Constants.OLECMDERR_E_NOTSUPPORTED;
+        }
+
+        public int Exec(ref Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
+        {
+            if (pguidCmdGroup == ReSharperCommandGroup && nCmdID >= ResumeId && nCmdID <= ToggleSuspendId)
+            {
+                // Don't delay command processing to update resharper status
+                Task.Run(() => InvokeBelowInputPriority(UpdateReSharperEnableStatus));
+            }
+
+            // No matter the command, we never actually want to respond to it, so always return not supported. We're just monitoring.
+            return (int)OLE.Interop.Constants.OLECMDERR_E_NOTSUPPORTED;
+        }
+
+        private void OnModalStateChanged(object sender, StateChangedEventArgs args)
+        {
+            ThisCanBeCalledOnAnyThread();
+
+            // Only monitor for StateTransitionType.Exit. This will be fired when the shell is leaving a modal state, including
+            // Tools->Options being exited. This will fire more than just on Options close, but there's no harm from running an
+            // extra QueryStatus.
+            if (args.TransitionType == StateTransitionType.Exit)
+            {
+                InvokeBelowInputPriority(UpdateReSharperEnableStatus);
+            }
+        }
+
+        public void Dispose(bool disposing)
+        {
+            if (!_disposedValue)
+            {
+                if (disposing)
+                {
+                    if (_priorityCommandTargetCookie != VSConstants.VSCOOKIE_NIL)
```

Comment:
it is initialized as NIL and nobody ever change it?

Reply:
Oh - just realized.  VS Platform has a new API for doing InfoBars in 15.6 - should we do this change in the vs-deps branch and consume that?

Reply:
It's only ever changed when we register.

---
In reply to: [155669412](https://github.com/dotnet/roslyn/pull/23605#discussion_r155669412) [](ancestors = 155669412)

Reply:
ah. Register call accepts this cookie. sorry, I missed that part. still question on when that Register call failed, whether it changes the value to non-nil or not. if it doesn't you are leaking event handler and etc.